src/BoundaryConditions/fill_halo_regions.jl

```diff
@@ -118,10 +144,10 @@ const PBCT  = Union{PBC,  NTuple{<:Any, <:PBC}}
 const MCBCT = Union{MCBC, NTuple{<:Any, <:MCBC}}
 const DCBCT = Union{DCBC, NTuple{<:Any, <:DCBC}}
 
-# Distributed halos have to be filled for last in case of 
-# buffered communication. Hence, we always fill them last
-
-# The reasoning for filling Periodic after Flux, Value, Gradient 
+# Distributed halos have to be filled last because of
+# buffered communication. 
```

Comment:
This doesn't explain it to me completely. Why would buffered communication require halos to be filled last? What do buffers have to do with it? Is that point that you have to complete other halo filling tasks before communicating, if you want to avoid filling the halos _again_ after communication?

It seems like we could communicate first before filling halos, right?

Reply:
I will update the comment. If you want asynchronous communication there are two options:

1) Fill all local halos - fill the buffers and start the communication - perform any required computation - complete the communication.
2) Fill the buffers and start the communication - perform any required computation - complete the communication - Fill any local halos.

with the first option, nothing changes in the case of non-distributed fields so I went in that direction in the #3125.

Reply:
By "complete the communication" you mean "wait until communication finishes" right?

Reply:
completing the communication requires waiting until the communication is finished and then filling the relevant halos from the received buffer. In practice, we can think about it as waiting until the communication is finished (assuming that filling the relevant halos is part of the communication process).

Reply:
Okay so the first step is to initiate communication, and the second step is to fill halos from the buffer. The code is misleading, because the initiation of communication is currently called "fill halo regions!". But halo filling is actually occurring in the second step, right?

Reply:
in asynchronous communication `fill_halo_regions!(f; async = true)` yes, the communication is completed in the `synchronize_communication!(f)` otherwise everything is completed in the `fill_halo_regions!` function

Reply:
Perhaps `finish_fill_halo_regions!(f)` or something would be helpful for the second step

Reply:
And probably neesd a comment in the code to indicate that the halo regions are not actually filled into the second step in the case of a distributed computation

Reply:
We assume that the second step completes everything: i.e. wait for communication to complete and fill the halos
https://github.com/CliMA/Oceananigans.jl/blob/d0b7ec8f98c860ce49927e0a7214961d2f47fb75/src/DistributedComputations/distributed_fields.jl#L56-L77